### PR TITLE
[Merged by Bors] - feat(Multilinear/Topology): define topology on `ContinuousMultilinearMap`s

### PR DESCRIPTION
**Pull request by urkud**
Define `UniformSpace` instance that doesn't depend on the norm, replace uniformity in the `SeminormedAddCommGroup` instance.

---

- [x] depends on: #10776 
- [x] depends on: #10743 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by ADedecker — APPROVED** [review]
Thanks a lot, and sorry for the delay!

My main comment is about wether we should put a uniform structure on `ContinuousMultilinearMap 𝕜 E F` if we only have a topology on `F`, the rest is pretty minor.

File: Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+@[simp]
+lemma toUniofrmOnFun_toFun [TopologicalSpace F] (f : ContinuousMultilinearMap 𝕜 E F) :
+    UniformOnFun.toFun _ f.toUniformOnFun = f :=
+  rfl
+
+instance instUniformSpace [TopologicalSpace F] [TopologicalAddGroup F] :
+    UniformSpace (ContinuousMultilinearMap 𝕜 E F) :=
+  letI := TopologicalAddGroup.toUniformSpace F
+  .comap toUniformOnFun inferInstance
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean:40]: I diverged from `ContinuousLinearMap` on purpose (and I was going to submit a PR changing `ContinuousLinearMap`s later): we don't have to worry about diamonds here; moreover, if `F` has canonical topology but no canonical uniformity, then we still can get a `UniformSpace` instance for continuous maps. TBH, I have no examples of vector spaces with a canonical topology but no canonical uniformity.

File: Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+namespace ContinuousMultilinearMap
+
+variable {𝕜 ι : Type*} {E : ι → Type*} {F : Type*}
+  [NormedField 𝕜]
+  [∀ i, TopologicalSpace (E i)] [∀ i, AddCommGroup (E i)] [∀ i, Module 𝕜 (E i)]
+  [AddCommGroup F] [Module 𝕜 F]
+
+/-- An auxiliary definition used to define topology on `ContinuousMultilinearMap 𝕜 E F`. -/
+def toUniformOnFun [TopologicalSpace F] (f : ContinuousMultilinearMap 𝕜 E F) :
+    (∀ i, E i) →ᵤ[{s | IsVonNBounded 𝕜 s}] F :=
```
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean:29]: Other files about `ContinuousMultilinerMap`s use `∀ i, E i`.

File: Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (ADedecker) [reply Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean:40]: My concern is not diamond issues (as you said we don't have to worry about diamonds here), but rather about not having nice defeqs. I guess you could argue that the whole construction is too complicated to be able to rely on defeq, but still I would prefer if the uniformity on continuous linear maps was defined in terms of the uniformity on the codomain. If that's too annoying, I would rather have only the uniform setup (assuming uniform space directly).

File: Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean (hunk 2 above)
[the remark this answers is not in the document]
Comment (ADedecker) [reply Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean:29]: I think this is only a remain of the switch to Lean4, when we had no `Π` notation. If you prefer to do the change in all these files at once that can also wait for a later PR of course.

File: Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean (hunk 1 above)
[the remark this answers is not in the document]
Comment (urkud) [reply Mathlib/Topology/Algebra/Module/Multilinear/Topology.lean:40]: I'll do this tonight.

**Comment by ADedecker** [comment]
Thanks!

bors merge

**Pull request closed without merging**